File: nothing/notifclip.py
"""Notification listing and clipboard management for Nothing phones."""

import re

from .device import run
from .exceptions import AdbError
from .models import DeviceInfo

_MAX_NOTIFICATIONS = 50


# ── helpers ───────────────────────────────────────────────────────────────────

def _shell(serial: str, cmd: str) -> str:
    r = run(["adb", "-s", serial, "shell", cmd])
    return r.stdout.strip() if r.returncode == 0 else ""


def _get_sdk(serial: str) -> int:
    """Return the Android SDK level as an integer, or 0 if unavailable."""
    raw = _shell(serial, "getprop ro.build.version.sdk")
    try:
        return int(raw)
    except ValueError:
        return 0


# ── notification parsing ──────────────────────────────────────────────────────

def _parse_notifications(output: str) -> list[dict[str, str]]:
    """
    Parse 'dumpsys notification --noredact' output and return a list of dicts
    with keys: pkg, title, text.

    The dumpsys format (Android 13+) groups fields inside NotificationRecord
    blocks.  A simplified excerpt looks like:

        NotificationRecord(0x1a2b3c4d: pkg=com.example.app ...)
          ...
          pkg=com.example.app
          ...
          android.title=Some Title
          android.text=Notification text
    """
    notifications: list[dict[str, str]] = []

    # Split on NotificationRecord boundaries so each chunk belongs to one entry.
    blocks = re.split(r"(?=\bNotificationRecord\b)", output)

    for block in blocks:
        if "NotificationRecord" not in block:
            continue

        pkg = ""
        title = ""
        text = ""

        for line in block.splitlines():
            stripped = line.strip()

            # pkg= can appear on the NotificationRecord header line or on its own.
            if not pkg:
                m = re.search(r"\bpkg=([^\s,)]+)", stripped)
                if m:
                    pkg = m.group(1)

            # android.title= / android.text= are the standard extras keys.
            # Values have format: String (actual text) or SpannableString (actual text)
            if not title:
                m = re.match(r"android\.title=(?:\w+\s+)?\(?(.*?)\)?$", stripped)
                if m:
                    raw = m.group(1).strip()
                    # Strip wrapper: "String (text)" → "text"
                    inner = re.match(r"(?:String|SpannableString)\s+\((.+)\)$", stripped[len("android.title="):].strip())
                    title = inner.group(1).strip() if inner else raw

            if not text:
                m = re.match(r"android\.text=(?:\w+\s+)?\(?(.*?)\)?$", stripped)
                if m:
                    raw = m.group(1).strip()
                    inner = re.match(r"(?:String|SpannableString)\s+\((.+)\)$", stripped[len("android.text="):].strip())
                    text = inner.group(1).strip() if inner else raw

        # Only include records that have at least a package name.
        if pkg:
            # Treat "null" (from Android dumpsys) the same as empty
            if title in ("null", "null\r"):
                title = ""
            if text in ("null", "null\r"):
                text = ""
            notifications.append({
                "pkg": pkg,
                "title": title if title else "(no title)",
                "text": text if text else "(no text)",
            })

    return notifications


# ── public actions ────────────────────────────────────────────────────────────

def action_notifications(device: DeviceInfo, package: str | None) -> None:
    """List active notifications on the device."""
    r = run(["adb", "-s", device.serial, "shell", "dumpsys notification --noredact"])
    if r.returncode != 0:
        raise AdbError(f"dumpsys notification failed: {r.stderr.strip()}")

    all_notifications = _parse_notifications(r.stdout)

    # Apply optional package filter.
    if package:
        shown = [n for n in all_notifications if n["pkg"] == package]
    else:
        shown = all_notifications

    # Enforce the hard cap.
    capped = shown[:_MAX_NOTIFICATIONS]

    print(f"\n  Active Notifications \u2014 {device.model}\n")

    if not capped:
        if package:
            print(f"  No notifications found for package: {package}\n")
        else:
            print("  No active notifications.\n")
        return

    # Column widths.
    pkg_w   = max(len("Package"),               max(len(n["pkg"])   for n in capped))
    title_w = max(len("Title"),                 max(len(n["title"]) for n in capped))
    # Clamp widths so the table stays readable on normal terminals.
    pkg_w   = min(pkg_w,   40)
    title_w = min(title_w, 30)

    header = (
        f"  {'#':<4} "
        f"{'Package':<{pkg_w}}  "
        f"{'Title':<{title_w}}  "
        f"Text"
    )
    print(header)
    print("  " + "-" * (len(header) - 2))

    for idx, notif in enumerate(capped, start=1):
        pkg_col   = notif["pkg"][:pkg_w]
        title_col = notif["title"][:title_w]
        text_col  = notif["text"]
        print(
            f"  {idx:<4} "
            f"{pkg_col:<{pkg_w}}  "
            f"{title_col:<{title_w}}  "
            f"{text_col}"
        )

    print()

    # Footer summary.
    total = len(all_notifications)
    shown_count = len(capped)

    if package:
        footer = f"  Total: {total} notification(s) ({shown_count} shown, filtered by package: {package})"
    elif shown_count < total:
        footer = f"  Total: {total} notification(s) (showing first {shown_count})"
    else:
        footer = f"  Total: {total} notification(s)"

    print(footer)
    print()


def action_clipboard(device: DeviceInfo, text: str | None) -> None:
    """Read or set the device clipboard."""
    serial = device.serial
    sdk = _get_sdk(serial)

    print(f"\n  Clipboard \u2014 {device.model}\n")

    if text is not None:
        # ── SET clipboard ─────────────────────────────────────────────────────
        # `cmd clipboard set-primary` is available from Android 13 (SDK 33).
        # On older versions it may still be present as an undocumented command;
        # we try it regardless and report the result.
        r = run(["adb", "-s", serial, "shell", "cmd", "clipboard", "set-primary", text])

        if r.returncode == 0 and "error" not in r.stdout.lower() and "unknown" not in r.stdout.lower():
            print(f"  Clipboard set to: {text}\n")
        else:
            # Fallback: try the short form without 'set-primary'.
            r2 = run(["adb", "-s", serial, "shell", "cmd", "clipboard", "set", text])
            if r2.returncode == 0 and "error" not in r2.stdout.lower() and "unknown" not in r2.stdout.lower():
                print(f"  Clipboard set to: {text}\n")
            else:
                print(f"  [WARN] Could not set clipboard via adb shell cmd clipboard.")
                if sdk and sdk < 29:
                    print(f"         Device is running SDK {sdk} — cmd clipboard may not be available.")
                else:
                    print(f"         The shell user may lack permission on this build (SDK {sdk}).")
                print()

    else:
        # ── READ clipboard ────────────────────────────────────────────────────
        content: str | None = None

        # Attempt 1: cmd clipboard get  (works on some AOSP/Nothing builds)
        r1 = run(["adb", "-s", serial, "shell", "cmd", "clipboard", "get"])
        if (
            r1.returncode == 0
            and r1.stdout.strip()
            and "error" not in r1.stdout.lower()
            and "unknown" not in r1.stdout.lower()
        ):
            content = r1.stdout.strip()

        # Attempt 2: low-level binder call  (service call clipboard 2 i32 0)
        if content is None:
            r2 = run(["adb", "-s", serial, "shell", "service", "call", "clipboard", "2", "i32", "0"])
            if r2.returncode == 0 and r2.stdout.strip():
                # Output looks like: Result: Parcel(00000000 00000002 00310033 ...)
                # The payload after the first two 32-bit words is UTF-16LE text.
                m = re.search(r"Result:\s*Parcel\(([0-9a-fA-F\s]+)\)", r2.stdout)
                if m:
                    hex_tokens = m.group(1).split()
                    # Drop the first two status/length words.
                    data_tokens = hex_tokens[2:]
                    raw_bytes = bytearray()
                    for token in data_tokens:
                        # Each token is a little-endian 32-bit word (8 hex chars).
                        try:
                            word = int(token, 16)
                            raw_bytes += word.to_bytes(4, byteorder="little")
                        except ValueError:
                            pass
                    # Decode as UTF-16LE and strip null terminators.
                    try:
                        decoded = raw_bytes.decode("utf-16-le").rstrip("\x00").strip()
                        if decoded:
                            content = decoded
                    except (UnicodeDecodeError, ValueError):
                        pass

        if content is not None:
            print(f"  Content: {content}\n")
        else:
            # Android 10+ restricts background clipboard access.
            sdk_label = f"SDK {sdk}" if sdk else "SDK unknown"
            print(f"  [INFO] Clipboard read requires foreground app access on Android 10+ ({sdk_label}).")
            print(f"         Background reads via adb shell are blocked by the OS.")
            print(f"         Use --text \"content\" to set clipboard instead.\n")

File: go/internal/notifclip/notifclip.go
// Package notifclip provides notification listing and clipboard management.
package notifclip

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

const maxNotifications = 50

func shellStr(serial, cmd string) string {
	stdout, _, code := adb.Run([]string{"adb", "-s", serial, "shell", cmd})
	if code != 0 {
		return ""
	}
	return strings.TrimSpace(strings.TrimRight(stdout, "\r\n"))
}

func getSDK(serial string) int {
	raw := shellStr(serial, "getprop ro.build.version.sdk")
	var sdk int
	fmt.Sscanf(raw, "%d", &sdk)
	return sdk
}

type notification struct {
	pkg, title, text string
}

var notifRecordRe    = regexp.MustCompile(`(?m)\bNotificationRecord\b`)
var pkgRe            = regexp.MustCompile(`\bpkg=([^\s,)]+)`)
var androidTitleRe   = regexp.MustCompile(`^android\.title=`)
var androidTextRe    = regexp.MustCompile(`^android\.text=`)
var spanWrapRe       = regexp.MustCompile(`^(?:String|SpannableString)\s+\((.+)\)$`)

func parseNotifications(output string) []notification {
	var notifications []notification
	blocks := splitOnBoundary(output, "NotificationRecord")

	for _, block := range blocks {
		if !strings.Contains(block, "NotificationRecord") {
			continue
		}

		var pkg, title, text string

		for _, line := range strings.Split(block, "\n") {
			stripped := strings.TrimRight(strings.TrimSpace(line), "\r")

			if pkg == "" {
				if m := pkgRe.FindStringSubmatch(stripped); m != nil {
					pkg = m[1]
				}
			}

			if title == "" && androidTitleRe.MatchString(stripped) {
				raw := stripped[len("android.title="):]
				if m := spanWrapRe.FindStringSubmatch(strings.TrimSpace(raw)); m != nil {
					title = m[1]
				} else {
					title = strings.TrimSpace(raw)
				}
			}

			if text == "" && androidTextRe.MatchString(stripped) {
				raw := stripped[len("android.text="):]
				if m := spanWrapRe.FindStringSubmatch(strings.TrimSpace(raw)); m != nil {
					text = m[1]
				} else {
					text = strings.TrimSpace(raw)
				}
			}
		}

		if pkg == "" {
			continue
		}
		if title == "null" || title == "null\r" {
			title = ""
		}
		if text == "null" || text == "null\r" {
			text = ""
		}
		if title == "" {
			title = "(no title)"
		}
		if text == "" {
			text = "(no text)"
		}
		notifications = append(notifications, notification{pkg: pkg, title: title, text: text})
	}
	return notifications
}

// splitOnBoundary splits text into sections before each occurrence of boundary word.
func splitOnBoundary(text, boundary string) []string {
	var blocks []string
	start := 0
	for i := 0; i < len(text); i++ {
		idx := strings.Index(text[i:], boundary)
		if idx == -1 {
			blocks = append(blocks, text[start:])
			break
		}
		absIdx := i + idx
		if absIdx > start {
			blocks = append(blocks, text[start:absIdx])
		}
		start = absIdx
		i = absIdx
	}
	if len(blocks) == 0 && start < len(text) {
		blocks = append(blocks, text[start:])
	}
	return blocks
}

// ActionNotifications lists active notifications on the device.
// packageName="" means show all.
func ActionNotifications(serial, model, packageName string) error {
	stdout, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell",
		"dumpsys notification --noredact"})
	if code != 0 {
		return nterrors.AdbError(fmt.Sprintf("dumpsys notification failed: %s", strings.TrimSpace(stderr)))
	}

	all := parseNotifications(stdout)

	shown := all
	if packageName != "" {
		var filtered []notification
		for _, n := range all {
			if n.pkg == packageName {
				filtered = append(filtered, n)
			}
		}
		shown = filtered
	}

	capped := shown
	if len(capped) > maxNotifications {
		capped = capped[:maxNotifications]
	}

	fmt.Printf("\n  Active Notifications \u2014 %s\n\n", model)

	if len(capped) == 0 {
		if packageName != "" {
			fmt.Printf("  No notifications found for package: %s\n\n", packageName)
		} else {
			fmt.Println("  No active notifications.\n")
		}
		return nil
	}

	pkgW := len("Package")
	titleW := len("Title")
	for _, n := range capped {
		if len(n.pkg) > pkgW {
			pkgW = len(n.pkg)
		}
		if len(n.title) > titleW {
			titleW = len(n.title)
		}
	}
	if pkgW > 40 {
		pkgW = 40
	}
	if titleW > 30 {
		titleW = 30
	}

	headerFmt := fmt.Sprintf("  %%-4s %%-%ds  %%-%ds  %%s", pkgW, titleW)
	header := fmt.Sprintf(headerFmt, "#", "Package", "Title", "Text")
	fmt.Println(header)
	fmt.Println("  " + strings.Repeat("-", len(header)-2))

	rowFmt := fmt.Sprintf("  %%-4d %%-%ds  %%-%ds  %%s", pkgW, titleW)
	for idx, notif := range capped {
		pkgCol := notif.pkg
		if len(pkgCol) > pkgW {
			pkgCol = pkgCol[:pkgW]
		}
		titleCol := notif.title
		if len(titleCol) > titleW {
			titleCol = titleCol[:titleW]
		}
		fmt.Printf(rowFmt+"\n", idx+1, pkgCol, titleCol, notif.text)
	}
	fmt.Println()

	total := len(all)
	shownCount := len(capped)

	var footer string
	switch {
	case packageName != "":
		footer = fmt.Sprintf("  Total: %d notification(s) (%d shown, filtered by package: %s)",
			total, shownCount, packageName)
	case shownCount < total:
		footer = fmt.Sprintf("  Total: %d notification(s) (showing first %d)", total, shownCount)
	default:
		footer = fmt.Sprintf("  Total: %d notification(s)", total)
	}
	fmt.Println(footer)
	fmt.Println()
	return nil
}

// ActionClipboard reads or sets the device clipboard.
// text="" means read mode; non-empty means set mode.
func ActionClipboard(serial, model, text string) error {
	sdk := getSDK(serial)

	fmt.Printf("\n  Clipboard \u2014 %s\n\n", model)

	if text != "" {
		out, _, code := adb.Run([]string{"adb", "-s", serial, "shell",
			"cmd", "clipboard", "set-primary", text})
		outLower := strings.ToLower(out)
		if code == 0 && !strings.Contains(outLower, "error") && !strings.Contains(outLower, "unknown") {
			fmt.Printf("  Clipboard set to: %s\n\n", text)
			return nil
		}
		// Fallback
		out2, _, code2 := adb.Run([]string{"adb", "-s", serial, "shell",
			"cmd", "clipboard", "set", text})
		out2Lower := strings.ToLower(out2)
		if code2 == 0 && !strings.Contains(out2Lower, "error") && !strings.Contains(out2Lower, "unknown") {
			fmt.Printf("  Clipboard set to: %s\n\n", text)
			return nil
		}
		fmt.Println("  [WARN] Could not set clipboard via adb shell cmd clipboard.")
		if sdk > 0 && sdk < 29 {
			fmt.Printf("         Device is running SDK %d — cmd clipboard may not be available.\n", sdk)
		} else {
			fmt.Printf("         The shell user may lack permission on this build (SDK %d).\n", sdk)
		}
		fmt.Println()
		return nil
	}

	// Read mode
	// Attempt 1: cmd clipboard get
	r1, _, c1 := adb.Run([]string{"adb", "-s", serial, "shell", "cmd", "clipboard", "get"})
	r1Lower := strings.ToLower(r1)
	if c1 == 0 && strings.TrimSpace(r1) != "" &&
		!strings.Contains(r1Lower, "error") && !strings.Contains(r1Lower, "unknown") {
		fmt.Printf("  Content: %s\n\n", strings.TrimSpace(r1))
		return nil
	}

	// Attempt 2: service call clipboard 2 i32 0
	r2, _, c2 := adb.Run([]string{"adb", "-s", serial, "shell",
		"service", "call", "clipboard", "2", "i32", "0"})
	if c2 == 0 && strings.TrimSpace(r2) != "" {
		decoded := decodeParcelUTF16(r2)
		if decoded != "" {
			fmt.Printf("  Content: %s\n\n", decoded)
			return nil
		}
	}

	sdkLabel := "SDK unknown"
	if sdk > 0 {
		sdkLabel = fmt.Sprintf("SDK %d", sdk)
	}
	fmt.Printf("  [INFO] Clipboard read requires foreground app access on Android 10+ (%s).\n", sdkLabel)
	fmt.Println("         Background reads via adb shell are blocked by the OS.")
	fmt.Println("         Use --text \"content\" to set clipboard instead.\n")
	return nil
}

var parcelRe = regexp.MustCompile(`Result:\s*Parcel\(([0-9a-fA-F\s]+)\)`)

func decodeParcelUTF16(output string) string {
	m := parcelRe.FindStringSubmatch(output)
	if m == nil {
		return ""
	}
	tokens := strings.Fields(m[1])
	if len(tokens) < 2 {
		return ""
	}
	dataTokens := tokens[2:]
	raw := make([]byte, 0, len(dataTokens)*4)
	for _, token := range dataTokens {
		var word uint32
		if _, err := fmt.Sscanf(token, "%x", &word); err != nil {
			continue
		}
		raw = append(raw,
			byte(word),
			byte(word>>8),
			byte(word>>16),
			byte(word>>24))
	}
	// Decode UTF-16LE
	if len(raw)%2 != 0 {
		raw = raw[:len(raw)-1]
	}
	runes := make([]rune, 0, len(raw)/2)
	for i := 0; i+1 < len(raw); i += 2 {
		r := rune(uint16(raw[i]) | uint16(raw[i+1])<<8)
		if r == 0 {
			break
		}
		runes = append(runes, r)
	}
	return strings.TrimSpace(string(runes))
}
